sycl/doc/extensions/proposed/sycl_ext_intel_virtual_functions.asciidoc

```diff
@@ -0,0 +1,579 @@
+- if an object is constructed in device code on a device `A`, calling a virtual
+  member function for that object in host code, or on another device `B` has
+  undefined behavior;
+
+=== New properties
+
+Due to the indirect nature of virtual member functions, compiler may not be able
+to understand which exact virtual member function is being called (i.e. which
+class it belongs to).  Moreover, when code is distributed over several
+translation units, compiler may not be able to even see definitions of all
+virtual member functions which may be called in a kernel.
+
+Therefore, to provide a mechanism for an implementation to help detect virtual
+member functions, which are going to be used from kernels and enforce necessary
+restrictions on them, new compile-time-constant properties are proposed.
+
+[source,dpcpp]
+----
+namespace sycl::ext::oneapi::experimental {
+
+  struct indirectly_callable_key {
+    template <typename SetId>
+    using value_t = property_value<indirectly_callable_key, SetId>;
+  };
+
+  struct calls_indirectly_key {
+    template <typename First = void, typename... SetIds>
+    using value_t = property_value<calls_indirectly_key, First, SetIds...>;
+  };
+
+  template <typename SetId = void>
+  inline constexpr indirectly_callable_key::value_t<SetId> indirectly_callable;
+
+  template <typename First = void, typename... Rest>
+  inline constexpr calls_indirectly_key::value_t<First, Rest...>
+      calls_indirectly;
+
+  template <>
+  struct is_property_key<indirectly_callable_key> : std::true_type {};
+  template <> struct is_property_key<calls_indirectly_key> : std::true_type {};
+}
+----
+
+Before describing those properties those properties in more detail, a couple of
+new terms are introduced to simplify the extension specification:
+
+Set of virtual member functions:: a group of virtual member functions which are
+defined with `indirectly_callable` property with the same value of the property
+parameter `SetId`. For the simplicity, this will also be further refferred to as
+a _set_, or as a _set of virtual functions_.
+
+Kernel declares a use of a set of virtual member functions:: a kernel is
+considered to be declaring a use of a set of virtual member functions `SetIdA`
+when it is submitted with `calls_indirectly` property with `SetIdA` included
+into the property parameter `SetIds`. If `SetIdA` is not included into the
+property parameter `SetIds`, or if a kernel is submitted without the property,
+then it is *not* considered to be declaring a use of the set of virtual member
+functions.
+
+|===
+|Property|Description
+|`indirectly_callable`
+|The `indirectly_callable` property indicates that a virtual member function is
+a device function, thus making it available to be called from SYCL kernel and
+device functions. Should only be applied to virtual member functions and to do
+so, function-style `SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY` macro should
```

Comment:
Why are we using this instead of the mechanism defined in the kernel properties extension?  That would give:

```
SYCL_EXT_ONEAPI_FUNCTION_PROPERTY((sycl::indirectly_callable<SetId>))
```

...and would allow multiple function properties to be defined in one place.

Reply:
Won't the FE have to implement that sort of parsing to handle compile-time properties eventually, anyway?  I'm trying to work out if this is just going to impact this extension (and #11861) or if you think there's a fundamental issue with the `SYCL_EXT_ONEAPI_FUNCTION_PROPERTY` design.

Reply:
> Won't the FE have to implement that sort of parsing to handle compile-time properties eventually, anyway?

Well, we are trying to avoid this for as long as possible :)

If I understand correctly, compile-time properties were introduced to move away from attributes with an idea that less compiler changes would be needed to implement a new property. We are still not in a state where introducing a new property can be solely done in headers, but we are not too far away from that for simple properties which are expected to trigger generation of certain SPIR-V decoration and nothing more.

Front-end is not that modular as the rest of the toolchain and any modifications to it are more likely to be scattered around the whole codebase instead of being isolated into a single SYCL-specific component like it is done for middle-end part of the toolchain (new tools and compiler passes). This I think is one of reasons why we don't want to touch the front-end too often.

> if you think there's a fundamental issue with the `SYCL_EXT_ONEAPI_FUNCTION_PROPERTY` design.

I don't quite like properties on free functions: on the one hand we are trying to move away from attributes, but on the other hand we are anyway using them, but in a very restricting manner which results in those custom properties macro we are introducing for virtual functions and free function kernel extensions.

Properties as an intended library-only replacement for attributes look good and solid for kernels. For free functions, however, they do not look as much as a library-only thing and hence all the problems and conflicts we see: need to resort to preprocessor to attach properties and inability (or rather reluctance) to do FE semantic checking based on properties.

I wonder if it is possible to implement something based on trait specializations instead?

```c++
// SYCL headers:
template<auto &Func, typename SetId>
struct is_device_indirectly_callable {
private:
  using __annotated_func = __some_magic_builtin_to_attach_metadata(Func, "indirecly-callable", __builtin_sycl_unique_stable_name(SetId));
  int __check = { []() __attribute__((sycl_device)) {
      // we are turning a function into a device one
      // by calling it from another device function
      Func();
      // yeah, that's a problem to write a generic code which would make a call to an arbitrary (member) function
     } }, 42; // operator comma to disregard the lambda later
};

// User code:

struct Foo {
  virtual void bar() { /* ... */ }
};

struct set_A;

template <>
is_device_indireclty_callable<Foo::bar, set_A> {};
// specialization is not enough if there won't be an instantiation, unfortunately
```

Reply:
Hm.  I don't think we considered traits for the function properties, and I'm not sure why.  I'll make some time to stare at this a bit more and see if there's any way that we could use traits as a general replacement here.  It would definitely be more library-only-friendly, and the macro has a lot of hacks I don't like (e.g., the double parentheses workaround for templated properties).

Reply:
I think switching to traits would allow us to adopt similar syntax to the syntax for [embedding properties into a kernel](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_properties.asciidoc#embedding-properties-into-a-kernel):

```c++
namespace syclex = sycl::ext::oneapi::experimental;

// Default trait in the headers
namespace sycl::ext::oneapi::experimental {

template <auto& Function>
struct function_traits {
    auto get(properties_tag) {
        return properties{};
    }
};

}

...

// Declare some device function
void foo();

// Attach properties to the device function
template <>
struct syclex::function_traits<foo> {
    auto get(syclex::properties_tag) {
        return syclex::properties{syclex::sub_group_size<8>, syclex::indirectly_callable<>};
    }
};
```

If we go this route, I think it makes more sense to attach all the properties to the function at once rather than to have separate traits (like `is_indirectly_callable` for each property).  This will lead to fewer surprises in the long run.

There are a few ways we could make the syntax above a little simpler, but there'd be some downsides (like functions and kernels requiring different syntax).

```c++
// Pass properties directly (shown as a property list but could be a parameter pack)
// NB: Renamed to _properties because we can't easily extend this version with other traits
template <>
struct syclex::function_properties<foo,
    decltype(syclex::properties{syclex::sub_group_size<8>, syclex::indirectly_callable<>})>
{};
```

```c++
// Use some sort of helper to define the right properties interface inside function_traits
// We could still extend this with other traits
template <>
struct syclex::function_traits<foo>:
  syclex::helper<decltype(syclex::properties{syclex::sub_group_size<8>, sycl::indirectly_callable<>})>
{};
```

Tagging @rlandschulz and @gmlueck for comment too.

Reply:
I think the traits solution won't work for `indirectly_callable` because our primary use case is to decorate a virtual member function.  The trait takes a non-type template parameter that is the address of the function, but you can only do this for functions declared at namespace scope and functions declared `static` in class scope.  Virtual functions are never static, so you would not be able to specialize the trait for these functions.

See here for an example: https://godbolt.org/z/3cj1reGfr

Reply:
I really don't understand the non-type template parameter rules or virtual function rules enough to understand what we need to achieve here, but it may be worth noting that things compile if the trait takes a pointer to the member instead: https://godbolt.org/z/ffz5q3Wvf

Reply:
I didn't realize you could do that!

I'm still not wild about the syntax, though.  The trait would have to be specialized after the function declaration, so it moves the property far away from the function to which it applies.

Reply:
A pointer to a member function requires an object to obtain an address of an actual function since, in general, the dynamic type of the object is needed to resolve an address. It isn't clear to me from the discussion so far how the trait would be used. Since it isn't possible from library code to identify the calls that are actually made from within a kernel or other device function, no opportunity for diagnosing calls to functions with unmatched set IDs is possible without compiler support. If compiler support is required, then use of a trait would just increase complexity.

Reply:
Even though visually traits look better than macro, they do not look good at all with virtual member functions.

Also, as @tahonermann noted we will still need some compiler magic to do diagnostics and apply special handling for those functions - and I'm not sure if that would even be possible if those traits are never used by the runtime.

We have agreed that it is ok for the compiler to parse string literals passed to the properties internal attribute and therefore the doc was updated to use the generic properties interface in 531083a82ccb48a2142bfcac6f8a71bd24813047

I will mark this thread as resolved